neat_cache/lib/src/providers/redis.dart

```diff
@@ -0,0 +1,149 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:io' show SocketException;
+import 'package:dartis/dartis.dart';
+import '../../cache_provider.dart';
+
+class _RedisContext {
+  final Connection connection;
+  final Client client;
+  final Commands<String, List<int>> commands;
+  _RedisContext({this.connection, this.client, this.commands});
+}
+
+typedef _Op<T> = Future<T> Function(Commands<String, List<int>>);
+
+class RedisCacheProvider extends CacheProvider<List<int>> {
+  final String _connectionString;
+  final Duration _connectTimeLimit;
+  final Duration _commandTimeLimit;
+  final Duration _reconnectDelay;
+
+  Future<_RedisContext> _context;
+  bool _isClosed = false;
+
+  RedisCacheProvider(
+    this._connectionString, {
+    Duration connectTimeLimit = const Duration(seconds: 30),
+    Duration commandTimeLimit = const Duration(milliseconds: 200),
+    Duration reconnectDelay = const Duration(seconds: 30),
+  })  : _connectTimeLimit = connectTimeLimit,
+        _commandTimeLimit = commandTimeLimit,
+        _reconnectDelay = reconnectDelay {
+    assert(_connectionString.isNotEmpty, 'connectionString must be given');
+    assert(!_connectTimeLimit.isNegative, 'connectTimeLimit is negative');
+    assert(!_commandTimeLimit.isNegative, 'commandTimeLimit is negative');
+    assert(!_reconnectDelay.isNegative, 'reconnectDelay is negative');
+  }
+
+  Future<_RedisContext> _createContext() async {
+    try {
+      print('### CONNECTING!!!');
+      final connection = await Connection.connect(_connectionString)
+          .timeout(_connectTimeLimit);
+      // Create context
+      final client = Client(connection);
+      return _RedisContext(
+        connection: connection,
+        client: client,
+        commands: client.asCommands(),
+      );
+    } on TimeoutException {
+      throw IntermittentCacheException('connect failed with timeout');
+    } on SocketException {
+      throw IntermittentCacheException('connect failed with socket exception');
+    } on Exception {
+      throw IntermittentCacheException('connect failed with exception');
+    }
+  }
+
+  Future<_RedisContext> _getContext() {
+    if (_context != null) {
+      return _context;
+    }
+    _context = _createContext();
+    scheduleMicrotask(() async {
```

Comment:
Do you need this microtask here?

Reply:
Yes, I'm returning before I'm awaiting...

1. If `_context` doesn't exist, I create it.
2. Then I return the `_context` future.
3. If I created `_context` in step (1), I then await it's failure, so that I can reset `_context = null` after a delay (if needed).

This ensures that we only make one connection at the time... no matter how many calls we have to `_getContext`.
Moreover, if the connection fails, we only do `_context = null` once. Until that has been done, all calls to `_getContext` will get a future that fails or a connection that is broken..

Reply:
Wondering, not sure: would this be better if you'd use a `Completer`? The logic could be replace with checking the completer if it exists (if it does return its `.future`, if it does not, trigger the creation of the context).

Reply:
```dart
  Completer<_RedisContext> _completer;
  Future<_RedisContext> _getContext() {
    if (_completer != null) {
      return _completer.future;
    }
    _completer = Completer<_RedisContext>();
    scheduleMicrotask(() async {
      _RedisContext ctx;
      try {
        ctx = await _createContext();
        _completer.complete(ctx);
      } on IntermittentCacheException {
        _completer.completeError(IntermittentCacheException);
        // If connecting fails, then we sleep and try again
        await Future.delayed(_reconnectDelay);
        _completer = null; // reset _completer, so next operation creates a new
        return;
      } catch (e) {
        // This is not an IntermittentCacheException, we can't retry, we should
        // propagate the error, hopefully, this will crash the application.
        _completer.completeError(e);
      }
      // If connecting was successful, then we await the connection being
      // closed or error, and we reset _completer.
      // Any errors have already been returned if any commands were affect
      // otherwise the error is just a broken TCP connection and we ignore it.
      try {
        await ctx.connection.done;
      } catch (e) {
        // ignore error
      }
      _completer = null;
    });
    return _completer.future;
  }
```

This is the completer variant, it just feels more fragile to it that way... but it highlights some behavior, so we can consider if it's what we want..

Reply:
Let's keep this current one.